Mathlib/Combinatorics/SimpleGraph/LapMatrix.lean

```diff
@@ -0,0 +1,331 @@
+  unfold mulVec dotProduct
+  simp only [Pi.one_apply, mul_one]
+  unfold degMatrix diagonal
+  simp only [of_apply, sum_ite_eq, mem_univ, ite_true, sub_self]
+
+theorem dotProduct_mulVec_adjMatrix [Ring α] (x : V → α) :
+    x ⬝ᵥ mulVec (G.adjMatrix α) x = ∑ i : V, ∑ j : V, if G.Adj i j then x i * x j else 0 := by
+  unfold dotProduct mulVec dotProduct
+  simp [mul_sum]
+
+theorem dotProduct_mulVec_degMatrix [CommRing α] (x : V → α) :
+    x ⬝ᵥ mulVec (G.degMatrix α) x = ∑ i : V, G.degree i * x i * x i := by
+  unfold dotProduct degMatrix
+  simp only [mulVec_diagonal, ← mul_assoc, mul_comm (x _)]
+
+variable (α)
+
+theorem sum_adj_eq_degree [AddCommMonoidWithOne α] (i : V) :
+    (G.degree i : α) = ∑ j : V, if G.Adj i j then 1 else 0 := by
+  rw [sum_boole]
+  unfold degree neighborFinset neighborSet
+  rw [Set.toFinset_setOf fun x ↦ Adj G i x]
+
+theorem ite_sub_ite {α : Type*} [Sub α] (P : Prop) [Decidable P] (a b c d : α) :
+    ((if P then a else b) - if P then c else d) = if P then a - c else b - d := by
+  split
+  repeat rfl
+
+theorem ite_add_ite {α : Type*} [Add α] (P : Prop) [Decidable P] (a b c d : α) :
+    ((if P then a else b) + if P then c else d) = if P then a + c else b + d := by
+  split
+  repeat rfl
+
+/-- Let $L$ be the graph Laplacian and let $x \in \mathbb{R}$, then
+ $$x^{\top} L x = \sum_{i \sim j} (x_{i}-x_{j})^{2}$$
+ where $\sim$ denotes the adjacency relation -/
+theorem lapMatrix_toLinearMap₂' (x : V → ℝ) : toLinearMap₂' (G.lapMatrix ℝ) x x =
+    (∑ i : V, ∑ j : V, if G.Adj i j then (x i - x j)^2 else 0) / 2 := by
+  rw [toLinearMap₂'_apply']
+  unfold lapMatrix
+  rw [sub_mulVec]
+  simp only [dotProduct_sub]
+  rw [dotProduct_mulVec_degMatrix, dotProduct_mulVec_adjMatrix, ← sum_sub_distrib]
+  simp only [sum_adj_eq_degree, sum_mul, ← sum_sub_distrib, ite_mul, one_mul,
+             zero_mul, ite_sub_ite, sub_zero]
+  rw [←half_add_self (∑ x_1 : V, ∑ x_2 : V, _)]
+  conv_lhs => arg 1; arg 2; arg 2; intro i; arg 2; intro j; rw [if_congr (adj_comm G i j) rfl rfl]
+  conv_lhs => arg 1; arg 2; rw [Finset.sum_comm]
+  simp [← sum_add_distrib]
+  conv_lhs => arg 1; arg 2; intro i; arg 2; intro j; rw [ite_add_ite]
+  field_simp
+  rw [sum_congr]
+  · rfl
+  · intros i _
+    rw [sum_congr]
+    · rfl
+    · intros j _
+      split
+      · rw [pow_two]
+        ring
+      · rfl
+
+/-- The Laplacian matrix is positive semidefinite -/
+theorem isPosSemidef_lapMatrix : (G.lapMatrix ℝ).PosSemidef := by
+  unfold PosSemidef
+  constructor
+  · unfold IsHermitian; rw [conjTranspose_eq_transpose_of_trivial, isSymm_lapMatrix]
+  · intro x
+    rw [star_trivial, ← toLinearMap₂'_apply', lapMatrix_toLinearMap₂']
+    apply div_nonneg_iff.mpr; left; constructor
+    · apply sum_nonneg'
+      intro i
+      apply sum_nonneg'
+      intro j
+      split_ifs
+      · exact sq_nonneg (x i - x j)
+      · rfl
+    · exact zero_le_two
+
+/-- `sqrtDiagMatrix A` is the diagonal matrix consisting of the square root of the diagonal
```

Comment:
All the lemmas about positive-semidef matrices will have to be moved into `Mathlib.LinearAlgebra.Matrix.PosDef` or somewhere similar. It might be nice to add, at the same time, the following characterisation of positive semidefinite matrices:
```
lemma Matrix.posSemidef_iff_exists_eq_transpose_mul_self {A : Matrix V V ℝ} : 
    PosSemidef A ↔ ∃ (B : Matrix V V ℝ), A = Bᵀ * B := 
  ⟨fun hA ↦ ⟨_, hA.spectral_theorem⟩, fun ⟨B, hB⟩ ↦ (hB ▸ posSemidef_conjTranspose_mul_self B)⟩
```

Reply:
I've made a separate PR #8809 for the linear-algebra parts (crediting you as a contributor) – I hope that's OK with you.

Reply:
Thats fine with me, thanks!